caffe2/core/dispatch/TensorTypeIdRegistration.h

```diff
@@ -0,0 +1,98 @@
+#pragma once
+
+/**
+ * To register your own tensor types, do in a header file:
+ *   C10_DECLARE_TENSOR_TYPE(MY_TENSOR)
+ * and in one (!) cpp file:
+ *   C10_DEFINE_TENSOR_TYPE(MY_TENSOR)
+ * Both must be in the same namespace.
+ */
+
+#include "caffe2/core/dispatch/TensorTypeId.h"
+#include "caffe2/core/common.h"
+#include <atomic>
+#include "flat_hash_map/flat_hash_map.h"
```

Comment:
Kindly do a size check for mobile.

Also, since we are doing a flat index, can we use an array directly?

Reply:
You mean array instead of hashmap and do linear scanning on access? We probably could but I don't think it's going to be faster.

Reply:
Yeah, my thought about array was mainly because of code size considerations, and also to help avoiding the third party2 header distribution problem. Speed is probably going to be the same since flat hash map is pretty much like that.

Reply:
We need flat_hash_map (or a different good hash map, i.e. no std::unordered_map) anyhow for the dispatch table, so using it here as well shouldn't increase binary size.

If flat_hash_map turns out to have too large a binary size, we can consider using the F14 ones instead.